File: Sources/MoonSchemaCompiler/Compile.swift
import MoonAST

public func compileSchemas(_ program: Program) -> CompileResult {
    var schemas: [String: MoonJsonSchema] = [:]
    var warnings: [SchemaWarning] = []

    for decl in program.declarations {
        switch decl {
        case .model(let modelDecl, _):
            schemas[modelDecl.name] = compileModel(modelDecl, warnings: &warnings)
        case .data(let dataDecl, _):
            compileData(dataDecl, schemas: &schemas, warnings: &warnings)
        default:
            break
        }
    }

    return CompileResult(schemas: schemas, warnings: warnings)
}

private func compileData(_ decl: DataDecl, schemas: inout [String: MoonJsonSchema], warnings: inout [SchemaWarning]) {
    let nullary = decl.constructors.filter { ctor in
        guard let args = ctor.args else { return true }
        if case .positional(let types, _) = args { return types.isEmpty }
        return false
    }

    let allNullaryOrPositional = decl.constructors.allSatisfy { ctor in
        guard let args = ctor.args else { return true }
        if case .record = args { return false }
        return true
    }

    if !nullary.isEmpty && allNullaryOrPositional {
        schemas[decl.name] = MoonJsonSchema(enumValues: nullary.map(\.name))
    }

    for ctor in decl.constructors {
        if case .record(let fields, _) = ctor.args {
            schemas[ctor.name] = compileRecordFields(
                ctor.name,
                fields: fields.map { RecordFieldSpec(name: $0.name, type: $0.type, modifiers: []) },
                warnings: &warnings
            )
        }
    }
}

private struct RecordFieldSpec {
    var name: String
    var type: TypeSpec
    var modifiers: [FieldModifier]
}

private func compileModel(_ decl: ModelDecl, warnings: inout [SchemaWarning]) -> MoonJsonSchema {
    compileRecordFields(
        decl.name,
        fields: decl.fields.map { RecordFieldSpec(name: $0.name, type: $0.type, modifiers: $0.modifiers) },
        warnings: &warnings
    )
}

private func compileRecordFields(
    _ name: String,
    fields: [RecordFieldSpec],
    warnings: inout [SchemaWarning]
) -> MoonJsonSchema {
    var properties: [String: MoonJsonSchema] = [:]
    var required: [String] = []

    for field in fields {
        let optional = field.modifiers.contains { if case .optional = $0 { return true }; return false }
        var schema = compileType(field.type)

        for mod in field.modifiers {
            if case .constraint(let expr, let span) = mod {
                schema = applyConstraint(schema, expr: expr, warnings: &warnings, line: span.start.line, column: span.start.column)
            }
        }

        properties[field.name] = schema
        if !optional {
            required.append(field.name)
        }
    }

    return MoonJsonSchema(
        type: "object",
        properties: properties,
        required: required.isEmpty ? nil : required,
        description: name
    )
}

private func compileType(_ spec: TypeSpec) -> MoonJsonSchema {
    switch spec {
    case .varType(let name, _):
        return typeNameToSchema(name)
    case .list(let element, _):
        return MoonJsonSchema(type: "array", items: compileType(element))
    case .con(let name, let args, _):
        if args.isEmpty {
            return typeNameToSchema(name)
        }
        return MoonJsonSchema(
            type: "object",
            properties: [:],
            required: [],
            description: "\(name) \(args.map(typeLabel).joined(separator: " "))"
        )
    case .tuple(let elements, _):
        return MoonJsonSchema(
            type: "array",
            items: MoonJsonSchema(enumValues: elements.indices.map(String.init)),
            description: "tuple(\(elements.map(typeLabel).joined(separator: ", ")))"
        )
    case .arrow(let from, let to, _):
        return MoonJsonSchema(description: "\(typeLabel(from)) -> \(typeLabel(to))")
    }
}

private func typeLabel(_ spec: TypeSpec) -> String {
    switch spec {
    case .varType(let name, _):
        return name
    case .list(let element, _):
        return "[\(typeLabel(element))]"
    case .con(let name, let args, _):
        if args.isEmpty { return name }
        return "\(name) \(args.map(typeLabel).joined(separator: " "))"
    case .tuple(let elements, _):
        return "(\(elements.map(typeLabel).joined(separator: ", ")))"
    case .arrow(let from, let to, _):
        return "\(typeLabel(from)) -> \(typeLabel(to))"
    }
}

private func typeNameToSchema(_ name: String) -> MoonJsonSchema {
    switch name {
    case "String": return MoonJsonSchema(type: "string")
    case "Int": return MoonJsonSchema(type: "integer")
    case "Float": return MoonJsonSchema(type: "number")
    case "Bool": return MoonJsonSchema(type: "boolean")
    default: return MoonJsonSchema(description: name)
    }
}

private func applyConstraint(
    _ schema: MoonJsonSchema,
    expr: Expression,
    warnings: inout [SchemaWarning],
    line: Int,
    column: Int
) -> MoonJsonSchema {
    if let between = parseBetweenConstraint(expr) {
        var updated = schema
        updated.minimum = between.min
        updated.maximum = between.max
        return updated
    }

    if let length = parseLengthConstraint(expr) {
        var updated = schema
        if schema.type == "string" {
            updated.maxLength = length
        } else {
            updated.maxItems = length
        }
        return updated
    }

    warnings.append(SchemaWarning(
        message: "Unrecognized constraint: \(exprToString(expr))",
        line: line,
        column: column
    ))
    return schema
}

private func parseBetweenConstraint(_ expr: Expression) -> (min: Double, max: Double)? {
    guard case .app(let outerFunc, let outerArg, _) = expr,
          case .app(let innerFunc, let innerArg, _) = outerFunc,
          case .varRef(let name, _) = innerFunc, name == "between",
          let min = literalNumber(innerArg),
          let max = literalNumber(outerArg) else {
        return nil
    }
    return (min, max)
}

private func parseLengthConstraint(_ expr: Expression) -> Int? {
    guard case .infix(let op, let left, let right, _) = expr, op == "<=",
          case .varRef(let name, _) = left, name == "length",
          case .lit = right,
          let value = literalNumber(right) else {
        return nil
    }
    return Int(value)
}

private func literalNumber(_ expr: Expression) -> Double? {
    guard case .lit(let value, _) = expr else { return nil }
    switch value {
    case .int(let n, _): return Double(n)
    case .float(let f, _): return f
    default: return nil
    }
}

private func exprToString(_ expr: Expression) -> String {
    switch expr {
    case .varRef(let name, _):
        return name
    case .lit(let value, _):
        switch value {
        case .string(let s, _): return "\"\(s)\""
        case .int(let n, _): return "\(n)"
        case .float(let f, _): return "\(f)"
        case .bool(let b, _): return "\(b)"
        }
    case .app(let fn, let arg, _):
        return "\(exprToString(fn)) \(exprToString(arg))"
    case .infix(let op, let left, let right, _):
        return "\(exprToString(left)) \(op) \(exprToString(right))"
    default:
        return "Expr"
    }
}

File: legacy/packages/schema-compiler/src/compile.ts
import type {
  Constructor,
  DataDecl,
  Expression,
  FieldDef,
  ModelDecl,
  Program,
  TypeSpec,
} from "@moon/ast";

export type JsonSchema = {
  type?: string;
  properties?: Record<string, JsonSchema>;
  required?: string[];
  items?: JsonSchema;
  enum?: string[];
  minimum?: number;
  maximum?: number;
  maxLength?: number;
  maxItems?: number;
  description?: string;
};

export interface SchemaWarning {
  message: string;
  line: number;
  column: number;
}

export interface CompileResult {
  schemas: Record<string, JsonSchema>;
  warnings: SchemaWarning[];
}

export function compileSchemas(program: Program): CompileResult {
  const schemas: Record<string, JsonSchema> = {};
  const warnings: SchemaWarning[] = [];

  for (const decl of program.declarations) {
    if (decl.kind === "Model") {
      schemas[decl.decl.name] = compileModel(decl.decl, warnings);
    } else if (decl.kind === "Data") {
      compileData(decl.decl, schemas, warnings);
    }
  }

  return { schemas, warnings };
}

function compileData(decl: DataDecl, schemas: Record<string, JsonSchema>, warnings: SchemaWarning[]): void {
  const nullary = decl.constructors.filter((c) => !c.args || (c.args.kind === "Positional" && c.args.types.length === 0));
  if (nullary.length > 0 && decl.constructors.every((c) => !c.args || c.args.kind !== "Record")) {
    schemas[decl.name] = {
      enum: nullary.map((c) => c.name),
    };
  }

  for (const con of decl.constructors) {
    if (con.args?.kind === "Record") {
      schemas[con.name] = compileRecordFields(con.name, con.args.fields, warnings);
    }
  }
}

function compileModel(decl: ModelDecl, warnings: SchemaWarning[]): JsonSchema {
  return compileRecordFields(decl.name, decl.fields, warnings);
}

function compileRecordFields(
  name: string,
  fields: FieldDef[] | { name: string; type: TypeSpec; modifiers?: FieldDef["modifiers"] }[],
  warnings: SchemaWarning[],
): JsonSchema {
  const properties: Record<string, JsonSchema> = {};
  const required: string[] = [];

  for (const field of fields) {
    const modifiers = "modifiers" in field ? field.modifiers : [];
    const optional = modifiers.some((m) => m.kind === "Optional");
    let schema = compileType(field.type);

    for (const mod of modifiers) {
      if (mod.kind === "Constraint") {
        schema = applyConstraint(schema, mod.expr, warnings, mod.span.start.line, mod.span.start.column);
      }
    }

    properties[field.name] = schema;
    if (!optional) {
      required.push(field.name);
    }
  }

  return {
    type: "object",
    description: name,
    properties,
    required,
  };
}

function compileType(spec: TypeSpec): JsonSchema {
  switch (spec.kind) {
    case "Var":
      return typeNameToSchema(spec.name);
    case "List":
      return { type: "array", items: compileType(spec.element) };
    case "Con":
      if (spec.args.length === 0) {
        return typeNameToSchema(spec.name);
      }
      return {
        type: "object",
        description: `${spec.name} ${spec.args.map((a) => typeLabel(a)).join(" ")}`,
        properties: {},
        required: [],
      };
    case "Tuple":
      return {
        type: "array",
        items: { enum: spec.elements.map((_, i) => String(i)) },
        description: `tuple(${spec.elements.map(typeLabel).join(", ")})`,
      };
    case "Arrow":
      return { description: `${typeLabel(spec.from)} -> ${typeLabel(spec.to)}` };
  }
}

function typeLabel(spec: TypeSpec): string {
  switch (spec.kind) {
    case "Var":
      return spec.name;
    case "List":
      return `[${typeLabel(spec.element)}]`;
    case "Con":
      return spec.args.length === 0 ? spec.name : `${spec.name} ${spec.args.map(typeLabel).join(" ")}`;
    case "Tuple":
      return `(${spec.elements.map(typeLabel).join(", ")})`;
    case "Arrow":
      return `${typeLabel(spec.from)} -> ${typeLabel(spec.to)}`;
  }
}

function typeNameToSchema(name: string): JsonSchema {
  switch (name) {
    case "String":
      return { type: "string" };
    case "Int":
      return { type: "integer" };
    case "Float":
      return { type: "number" };
    case "Bool":
      return { type: "boolean" };
    default:
      return { description: name };
  }
}

function applyConstraint(
  schema: JsonSchema,
  expr: Expression,
  warnings: SchemaWarning[],
  line: number,
  column: number,
): JsonSchema {
  const between = parseBetweenConstraint(expr);
  if (between) {
    return { ...schema, minimum: between.min, maximum: between.max };
  }

  const length = parseLengthConstraint(expr);
  if (length !== null) {
    if (schema.type === "string") {
      return { ...schema, maxLength: length };
    }
    if (schema.type === "array") {
      return { ...schema, maxItems: length };
    }
    return { ...schema, maxItems: length };
  }

  warnings.push({ message: `Unrecognized constraint: ${exprToString(expr)}`, line, column });
  return schema;
}

function parseBetweenConstraint(expr: Expression): { min: number; max: number } | null {
  if (expr.kind !== "App" || expr.arg.kind !== "Lit") return null;
  const inner = expr.func;
  if (inner.kind !== "App" || inner.func.kind !== "Var" || inner.func.name !== "between") return null;
  if (inner.arg.kind !== "Lit") return null;
  const min = literalNumber(inner.arg);
  const max = literalNumber(expr.arg);
  if (min === null || max === null) return null;
  return { min, max };
}

function parseLengthConstraint(expr: Expression): number | null {
  if (expr.kind !== "Infix" || expr.op !== "<=") return null;
  if (expr.left.kind !== "Var" || expr.left.name !== "length") return null;
  if (expr.right.kind !== "Lit") return null;
  return literalNumber(expr.right);
}

function literalNumber(expr: Expression): number | null {
  if (expr.kind !== "Lit") return null;
  if (expr.value.kind === "Int" || expr.value.kind === "Float") return expr.value.value;
  return null;
}

function exprToString(expr: Expression): string {
  switch (expr.kind) {
    case "Var":
      return expr.name;
    case "Lit":
      return JSON.stringify(expr.value.kind === "String" ? expr.value.value : expr.value.value);
    case "App":
      return `${exprToString(expr.func)} ${exprToString(expr.arg)}`;
    case "Infix":
      return `${exprToString(expr.left)} ${expr.op} ${exprToString(expr.right)}`;
    default:
      return expr.kind;
  }
}

